sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubAuthenticationPolicy.cs

```diff
@@ -26,12 +25,31 @@ internal partial class WebPubSubAuthenticationPolicy : HttpPipelineSynchronousPo
         public override void OnSendingRequest(HttpMessage message)
         {
             string audience = message.Request.Uri.ToUri().AbsoluteUri;
-            var expiresAt = DateTime.UtcNow + TimeSpan.FromMinutes(10);
+            var now = DateTimeOffset.UtcNow;
+            var expiresAt = now + TimeSpan.FromMinutes(5);
 
-            string accessToken = JwtUtils.GenerateJwtBearer(audience, claims: null, expiresAt, _credential);
+            // TODO: is this a bug in the service?
+            // The service accepts ASCII of the portal key, but not Base64 decoded.
+            //
+            //var keyBytes = Convert.FromBase64String(_credential.Key);
+            var keyBytes = Encoding.UTF8.GetBytes(_credential.Key);
 
-            var header = new AuthenticationHeaderValue("Bearer", accessToken);
-            message.Request.Headers.SetValue(HttpHeader.Names.Authorization, header.ToString());
+            var writer = new JwtBuilder(keyBytes);
+            writer.AddClaim(JwtBuilder.Nbf, now);
+            writer.AddClaim(JwtBuilder.Exp, expiresAt);
+            writer.AddClaim(JwtBuilder.Iat, now);
+            writer.AddClaim(JwtBuilder.Aud, audience);
+            int jwtLength = writer.End();
+
+            var prefix = "Bearer ";
+            var state = (prefix, writer);
+            var headerValue = NS2Bridge.CreateString(jwtLength + prefix.Length, state, (destination, state) => {
```

Comment:
Just wondering - what's up with the `state` parameter to `NS2Bridge.CreateString?` couldn't the delegate just close over anything it wanted?

Is this something to do with the fact that `(prefix, writer`) allocates a ValueTuple on the stack and so we don't have a heap allocation for the display class?

Reply:
Yup. The closure is an allocation. BTW, this exact API exists in .NET 5.0 and the state is used to avoid allocations.